File: kDiffArray.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

import java.util.HashMap;

public class kDiffArray {
    public static void main(String args[]) {
        int[] nums = {3,1,4,1,5};
        int k = 2;
        // create a hash map to store the freuency count for the numbers in the array
        HashMap<Integer,Integer> map = new HashMap<Integer,Integer>();
        int count = 0;
        for(int i = 0; i < nums.length; i++) {
            if (!map.containsKey(nums[i]))
            {
                map.put(nums[i],1);
            }
            else {
                map.put(nums[i],map.get(nums[i]) + 1);
            }    
        }
        
        // if the key + k is present which is y = k + x. So for difference we see the addition part to find the other number present
        // in the hashmap
        if (k > 0) {
        for(int key: map.keySet()) 
        {
            if(map.containsKey(key + k))
                count++;     
        }
        }
        else {
            // edge case if k is zero then return the number with one or more duplicates to form a unique pair.
            for(int key: map.keySet()) 
        {
            if(map.containsKey(key + k) && map.get(key + k) >=2)
                count += 1;     
        }
        }
        System.out.println(count);
    }
    
}

File: K Diff Array.py
# Time Complexity : O(n)
# Space Complexity : O(n) to maintain the hash map

# The code ran on LeetCode

# Maintain a count of all the visited until now in a hash map. At every element in the nums array, check if the left pair or right pair exists. Add all the unique pairs to the res array. 
# Handle the case k = 0 separately. At most, an element can be added to the res only once in this case.
# Return the length of the res array

class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        res = []
        d = {}
        for n in nums:
            if n not in d:
                # Check for left pair:
                if n-k in d and [n-k, n] not in res:
                    res.append([n-k, n])
                # Check for right pair
                if n+k in d and [n, n+k] not in res:
                    res.append([n, n+k])
                d[n] = 1
            else:
                d[n]+=1
                if k == 0 and d[n] >= 2 and [n, n] not in res:
                    res.append([n, n])
        return len(res)